src/blocks/mrc_mechanism.ts

```diff
@@ -0,0 +1,143 @@
+  type: string,
+};
+
+type MechanismExtraState = {
+  importModule?: string,
+  params?: ConstructorArg[],
+}
+
+type MechanismBlock = Blockly.Block & MechanismMixin;
+interface MechanismMixin extends MechanismMixinType {
+  mrcArgs: ConstructorArg[],
+  mrcImportModule: string,
+
+}
+type MechanismMixinType = typeof MECHANISM_FUNCTION;
+
+const MECHANISM_FUNCTION = {
+  /**
+    * Block initialization.
+    */
+  init: function (this: MechanismBlock): void {
+    this.setStyle(MRC_STYLE_FUNCTIONS);
+    this.appendDummyInput()
+      .appendField(new Blockly.FieldTextInput('my_mech'), 'NAME')
+      .appendField('of type')
+      .appendField(createFieldNonEditableText(''), 'TYPE');
+    this.setPreviousStatement(true);
+    this.setNextStatement(true);
+  },
+
+  /**
+    * Returns the state of this block as a JSON serializable object.
+    */
+  saveExtraState: function (this: MechanismBlock): MechanismExtraState {
+    let extraState: MechanismExtraState = {
+    };
+    extraState.params = [];
+    this.mrcArgs.forEach((arg) => {
+      extraState.params!.push({
+        'name': arg.name,
+        'type': arg.type,
+      });
+    });    
+    if (this.mrcImportModule) {
+      extraState.importModule = this.mrcImportModule;
+    }
+    return extraState;
+  },
+  /**
+  * Applies the given state to this block.
+  */
+  loadExtraState: function (this: MechanismBlock, extraState: MechanismExtraState): void {
+    this.mrcImportModule = extraState.importModule ? extraState.importModule : '';
+    this.mrcArgs = [];
+
+    if(extraState.params){
+      extraState.params.forEach((arg) => {
+        this.mrcArgs.push({
+          'name': arg.name,
+          'type': arg.type,
+        });
+      });
+    }
+    this.mrcArgs = extraState.params ? extraState.params : [];
+    this.updateBlock_();
+  },
+  /**
+     * Update the block to reflect the newly loaded extra state.
+     */
+    updateBlock_: function(this: MechanismBlock): void {
+      // Add input sockets for the arguments.
+      for (let i = 0; i < this.mrcArgs.length; i++) {
+        const input = this.appendValueInput('ARG' + i)
+            .setAlign(Blockly.inputs.Align.RIGHT)
+            .appendField(this.mrcArgs[i].name);
+        if (this.mrcArgs[i].type) {
+          input.setCheck(getAllowedTypesForSetCheck(this.mrcArgs[i].type));
+        }
+      }
+    }
+}
+
+export const setup = function () {
+  Blockly.Blocks[BLOCK_NAME] = MECHANISM_FUNCTION;
+}
+
+//TODO: This needs to cause our own init to create the mechanisms line
+export const pythonFromBlock = function (
+  mechanismBlock: MechanismBlock,
+  generator: ExtendedPythonGenerator,
+) {
+  if (mechanismBlock.mrcImportModule) {
+    generator.addImport(mechanismBlock.mrcImportModule);
+  }
+  let code = 'self.mechanisms["' + mechanismBlock.getFieldValue('NAME') + '"] = '
+    + mechanismBlock.getFieldValue('TYPE') + '('
+
+  for (let i = 0; i < mechanismBlock.mrcArgs.length; i++) {
+      let fieldName = 'ARG' + i;
```

Comment:
let -> const

Reply:
Sure.  I copied it from your mrc_call_python (Line 245) so we might want to change it there as well.

Reply:
Oops, I misunderstood which let you were talking about.